profiling/src/lib.rs

```diff
@@ -554,25 +524,44 @@ fn log_add_tag(locals: &mut RefMut<RequestLocals>, key: &str, value: &str) {
     }
 }
 
-fn detect_url_from_globals(globals: &DatadogPhpProfilingGlobals) -> anyhow::Result<Uri> {
-    if globals.trace_agent_url.size > 0 {
-        let maybe_uri: &str = (&globals.trace_agent_url).try_into()?;
-        Ok(Uri::from_str(maybe_uri)?)
-    } else if globals.trace_agent_port.size > 0 || globals.agent_host.size > 0 {
-        let host: &str = if globals.agent_host.size > 0 {
-            (&globals.agent_host).try_into()?
-        } else {
-            "localhost"
-        };
-        let port: &str = if globals.trace_agent_port.size > 0 {
-            (&globals.trace_agent_port).try_into()?
-        } else {
-            "8126"
-        };
-        Ok(Uri::from_str(format!("http://{}:{}", host, port).as_str())?)
-    } else {
-        Ok(Uri::from_str("http://localhost:8126")?)
+fn detect_uri_from_env(env: &config::Env) -> AgentEndpoint {
+    /* Priority:
+     *  1. DD_TRACE_AGENT_URL
+     *     - RFC allows unix:///path/to/some/socket so parse these out.
+     *     - Maybe emit diagnostic if an invalid URL is detected or the path is non-existent, but
+     *       continue down the priority list.
+     *  2. DD_AGENT_HOST and/or DD_TRACE_AGENT_PORT. If only one is set, default the other.
+     *  3. Unix Domain Socket at /var/run/datadog/apm.socket
+     *  4. http://localhost:8126
+     */
+    if let Some(trace_agent_url) = &env.trace_agent_url {
+        // check for UDS first
+        if let Some(path) = trace_agent_url.strip_prefix("unix://") {
+            let path = PathBuf::from(path);
+            if path.exists() {
+                return AgentEndpoint::Socket(path);
+            }
+        }
+        match Uri::from_str(trace_agent_url) {
+            Ok(uri) => return AgentEndpoint::Uri(uri),
+            Err(err) => warn!("DD_TRACE_AGENT_URL was not a valid URL: {}", err),
+        }
```

Comment:
```suggestion
            } else {
                warn!("Unix socket specified in DD_TRACE_AGENT_URL does not exist: {} ", path);
            }
        } else {
            match Uri::from_str(trace_agent_url) {
                Ok(uri) => return AgentEndpoint::Uri(uri),
                Err(err) => warn!("DD_TRACE_AGENT_URL was not a valid URL: {}", err),
            }
        }
```
The URL is valid, but it doesn't exist, we should distinguish these cases.

Reply:
I'll look at this in a bit. Thanks!

Reply:
Looks good, but need to call `.to_string_lossy()` on the `path` since it doesn't implement `Display` (OS strings are not guaranteed to be UTF-8), so I won't commit this directly.

Reply:
The alternative would be renaming the path variable and use the actual string which you extracted from the tracer_agent_url string.